### PR TITLE
Assignment 3

### PR DESCRIPTION
**Pull request by VolhaAnishchanka**

### CODE REVIEW

File: HelgaShiryaeva/assignment3/DictionaryTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Set;
+import java.util.HashSet;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class DictionaryTest {
+    private Dictionary dictionary;
+    private WordSearch wordSearch;
```
Comment (brzozan) [inline HelgaShiryaeva/assignment3/DictionaryTest.java]: `WordSearch` should have its own test class. Idea behind unit tests is that they test a single unit of code, which in case of Java is typically a single class.

File: HelgaShiryaeva/assignment3/DictionaryTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @Test
+    public void isPrefixTest() {
+        assertTrue(dictionary.isPrefix("c"));
+        assertTrue(dictionary.isPrefix("ca"));
+        assertTrue(dictionary.isPrefix("cat"));
+        assertFalse(dictionary.isPrefix("fdkl"));
+    }
+
+    @Test
+    public void wordSearchTest() {
```
Comment (brzozan) [inline HelgaShiryaeva/assignment3/DictionaryTest.java]: It would be good to verify edge cases too.

File: HelgaShiryaeva/assignment3/DictionaryTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        char[][] grid = new char[2][3];
+        grid[0][0] = 'a'; grid[0][1] = 'a'; grid[0][2] = 'r';
```
Comment (brzozan) [inline HelgaShiryaeva/assignment3/DictionaryTest.java]: Simpler version:
```
char[][] grid = new char[][] {
  {'a', 'a', 'r'},
  {'t', 'c', 'd'}
};
```

File: HelgaShiryaeva/assignment3/WordSearch.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        prefix.append(board[i][j]);
+        if(!dictionary.isPrefix(prefix.toString())) {
+            prefix.deleteCharAt(prefix.length() - 1);
+            return;
+        }
+        if(dictionary.isWord(prefix.toString())) {
+            words.add(prefix.toString());
+        }
+        isVisited[i][j] = true;
+        dfs(board, i, j - 1, prefix, words, isVisited);
```
Comment (brzozan) [inline HelgaShiryaeva/assignment3/WordSearch.java]: Code would be easier to understand if you used nested loops to iterate over offsets. With current list of almost-identical-but-not-really calls to dfs person reading code needs to mentally keep track of 8 sets of param values to make sure that correct fields are visited and it is easy to miss a bug if at one place there is a '+' instead of '-'.

**Review by brzozan — CHANGES_REQUESTED** [review]
Very, very clean and well commented PR. Thank you!

I added some relatively minor comments.

**Review by brzozan — APPROVED** [review]
LGTM, thank you!

**Comment by VolhaAnishchanka** [comment]
Thank you so much for the review ! :)
